docs/markdown/snippets/optionrefactor.md

```diff
@@ -0,0 +1,45 @@
+## Per project subproject options rewrite
+
+A common requirement when building large projects with many
+subprojects is to build some (or all) subprojects with project options
+that are different from the "main project". This has been sort of
+possible in a limited way but is now possible in a general way. These
+additions can be added, changed and removed at runtime using the
+command line or, in other words, without editing existing
+`meson.build` files.
+
+Since this release you can specify _augments_, which are basically
+per-project option settings. These can be specified for every top
+level (i.e. not project) options. Suppose you have a project that has
+a single subproject called `numbercruncher` that does heavy
+computation. During development you want to build that subproject with
+optimizations enabled but your main project without
+optimizations. This can be done by specifying an augment to the given
+subproject:
+
+    meson configure -Anumbercruncher:optimization=3
+
+Another case might be that you want to build with errors as warnings,
+but some subproject does not support it. It would be set up like this:
+
+    meson configure -Dwerror=true -Anaughty:werror=false
+
+You can also specify an augment on the opt level project. A more
+general version of enabling optimizations on all subprojects but not
+the top project would be done like this:
+
+    meson configure -Doptimization=2 -A:optimization=0
```

Comment:
I think this should be `    meson configure -Doptimization=0 -A:optimization=2` with the values passed to the `optimization` option swapped.

Reply:
Not really:

`-Doptimization=2 -A:optimization=0` means "optimize everything _except_ the top level project.

`-Doptimization=0 -A:optimization=2` means "optimize top project _only_".

The former is you want most of the time and what happens when you use Linux distro dependencies. The `-'A:bob` syntax means _apply only to the top level project_, not _apply to all subprojects_.

If this syntax is confusing we can change it. I did go through a bunch of them and this seemed to be the most consistant and simple, but if someone comes up with a better syntax, great.

Reply:
Uhm. I indeed misunderstood the syntax for the `-A` option argument. After thinking for a brief moment, I think the syntax implemented here makes sense. 

However, I'm not sure of the reason why both the `-A` and `-D` options are needed. I think that `meson configure -Doptimization=0 -A:optimization=2` could be written as `meson configure -Doptimization=0 -D:optimization=2` and that the argument to the second `-D` option could be parsed to make it into something that applies to the top project only (valid option names cannot contain columns, AFAIK). Having two syntax for specifying setup options feels strange. Similarly, the `-U` option would apply both to global options ans to project specific options.

Reply:
Why does the action of creating an augment need to be explicit, from the point of view of the user? I'm also not sure about how it combines with `meson setup --reconfigure`. Should `-A` or `-D` be used to set an augment when running `meson setup --reconfigure`? This is particularly confusing because recent meson versions accent the `--reconfigure` option also when the specified build directory is not yet initialized.

Currently there is no way to reset an option to the project default, thus it would be nice to be able to use the `-U` option to unset auments and global options.

Reply:
> Why does the action of creating an augment need to be explicit, from the point of view of the user?

Because having a thing that does two logically different things based on not-immediately-visible state is confusing. In this case `-D` would mean either "create a new subproject option" or "change the value of an existing option or create a new one if it does not exist yet".

> recent meson versions accent the --reconfigure option also when the specified build directory is not yet initialized.

IMHO that is a bug. Reconfigure should only work on existing build dirs and not have any extra options.

> it would be nice to be able to use the -U option to unset auments and global options.

Which would again mean that it does two logically different things "remove per-project setting" and "restore setting to its original value for this project (somehow"). It is also not clear what that value should be. If a user does `meson setup --prefix=/foo`, then does `meson configure -Dprefix=/bob` then does a reset, should the value go to `/foo`, `/bar/` or `/usr/local` (the actual default value of `prefix`)? 

The "original value" is currently not stored anywhere, so we could not restore it even if we wanted to.

Reply:
I don't understand why, from the point of view of the user, it is important to know that augments are created, set to a specific value, and removed. Doesn't the user care only about what value for a given project options is used?

To set options `bar` for subproject `foo` to the value 42, the user would always run `meson setup -Afoo:bar=42`. Once the build dir is setup, to change the value of the option, I would run `meson configure -Dfoo:bar=25`. However, for setting another option not previously set, I need `meson configure -Afoo:other=false`. I cannot think of an use case where getting an error because the augment has not been created previously would be useful.

> > recent meson versions accent the --reconfigure option also when the specified build directory is not yet initialized.
> 
> IMHO that is a bug. Reconfigure should only work on existing build dirs and not have any extra options.

Actually, this has been explicitly recently changed to behave like this.

> Which would again mean that it does two logically different things "remove per-project setting" and "restore setting to its original value for this project (somehow").

I still don't get it. The model I have for how options work (or should work) is that of layered mappings. In order of decreasing priority, the mappings are: target options, project options, global options. Target options are set in `meson.build`, project options are set via `project(default_options={})`, and global options are the setup options as they exist now. This PR interposes another layer between target options and project options that we call "augments". In my mind, the `-D` option only adds an entry in the global options dictionary, and the `-A` option adds an entry in the relevant project augments dictionary. The `-U` options would simply delete the specified entry from these dictionaries.

From the conversation we are having, I evince that this model is wrong. But I don't understand how.

Reply:
>  In my mind, the -D option only adds an entry in the global options dictionary,

No! Absolutely not! With this you can not add "project specific" things to the global dictionary. The parts in the code where this was possible has (or will be) converted to augments. There **must not under any circumstances** be two separate places where you can add per-subproject option values. That way lies madness.

Reply:
It maybe expressed myself not correctly in describing what I had in mind, but clearly you didn't do any effort in trying to understand what I was trying to say. You still haven't explained why, from the user point of view, creating a project specific option and setting it need to be two conceptually different thing. To me it is similar to require Python code written like
```
d = {}
d.create('foo', 1)
d.set('foo', 2)
```
to create a key-value pair in a dictionary and then set it do a different value. The only explanation provided is that these are different operations. While this may be true to some extent, the ergonomic of having two command line options is awful and as far as I can tell completely artificial. However, it is evident that you don't want to really discuss the design, so I'll shut up.

Reply:
Let me state that I did consider both the version where `-D` can be used to both add and set as well as having a separated flag, say `-AD` to do that (which is basically what SQL's `upsert` is). There have been a few cases in the past where we have added helper code for "let's just do this automatically as that is what people almost certainly mean" only to discover that it actually wasn't. Because of this I'm somewhat cautious on adding "do what I mean" functionality by default.

Another thing to note is that it is easy to take things that used to be hard errors and turn them into something usable. OTOH trying to take something that is permitted and trying to make it prohibited is either really hard or just plain impossible. Because of this is typically better to start with a simple, limited and rigid UI and loosen or add more convenient methods later than the other way around.

As a specific example, if `-D` were to automatically add an option if it did not already exist, then there would not be a way to say "add per-project option X but only if it does not already exist". How common is this requirement? I don't know. Probably not very. But we don't really know for sure until people start using this functionality.

Reply:
> As a specific example, if `-D` were to automatically add an option if it did not already exist, then there would not be a way to say "add per-project option X but only if it does not already exist".

I think this use case is quite far fetched, but I think it is the only one that justifies separating `-A` and `-D` into two separate options. Therefore, we should review how the proposed command line interface covers this use case. 

Assuming I have a setup build directory and I want to set option `bar` for sub-project `foo` only if it already exist, I would do 
```
meson configure -Dfoo:bar=42
```
This would result in an error if the augment `foo:bar` does not exist. However, how do I know that the command failed because the augment does not exist or for an unrelated reason? On the other hand, in the IMO much more common case where I want to set the same option regardless of whether it already exists or not, I need to execute 
```
meson configure -Dfoo:bar=42 || meson configure -Afoo:bar=42
```
which seems a bit too verbose.

I think this use case would be better served by a way to query the current value of a specific option or augment. Assuming that the `-Q` option returns the value of a specific option or augment, I could do
```
meson configure -Qfoo:bar || meson configure -Dfoo:bar=42
```
Where `meson configure -Qfoo:bar` exists with a specific status code different from zero when the option is not found. I think the possibility of querying the value of options is a nice introspection facility that would be nice to have.

Reply:
I haven't taken a good look at how this PR handles it all, but one straightforward way to define the distinction is by looking at something like `prefix`.

With this PR, what happens if you try to "augment" the prefix option?

With this PR, what option gets looked up when you do `get_option('prefix')` in a subproject? What is the user-facing mental model of that option? This scenario doesn't appear to be covered by the unittests, at least, and it should.

Reply:
I added a commit that removes -A in favour of -D. Docs not updates. Does this make the UI better?

Reply:
Ping @dnicolodi.

Reply:
Thanks for being receptive of my comments @jpakkane. I haven't had the occasion to delve into the implementation, and my previous comments were based almost entirely on reading the additions to the documentation, which is not yet updated here. However, having only the `-D` option clearly solves most of my issue with the new feature.